### PR TITLE
feat(Label): Added support for React router

### LINKED ISSUE #4811: support react-router Link with Label
**Describe the issue. What is the expected and unexpected behavior?**

We need a way to use the `Label` component with `react-router` `Link`. However the `Label` component does not support rendering the component as a `Link`.

**Is this a bug or enhancement? If this issue is a bug, is this issue blocking you or is there a work-around?**
Enhancement.

Workaround is to supply an `onClick` prop with `history.push(...)` 

**What is your product and what release version are you targeting?**
OpenShift 4.7

### PR DESCRIPTION
**Pull request by tlabaj**
Closes #4811 

cc@mcoker

### CODE REVIEW

File: packages/react-core/src/components/Label/Label.tsx [hunk 1]
```diff
@@ -79,11 +92,14 @@ export const Label: React.FunctionComponent<LabelProps> = ({
         className
       )}
     >
-      <Component className={css(styles.labelContent)} {...(href && { href })}>
-        {icon && <span className={css(styles.labelIcon)}>{icon}</span>}
-        {isTruncated && <span className={css(styles.labelText)}>{children}</span>}
-        {!isTruncated && children}
-      </Component>
+      {customContent ? (
+        <>{customContent}</>
```
Comment (mcoker) [inline packages/react-core/src/components/Label/Label.tsx:96]: are we able to add `css(styles.labelContent)` to whatever is passed in?

**Comment by jschuler** [comment]
@tlabaj I would use the render props pattern here. Basically it's just another prop (usually called `render`) that gives the user control over the returned content. I could see it being used like this:
```
const FakeRouterLink = ({ to, children, ...props }) => <a href={to} {...props}>{children}</a>;
...
<Label color="blue" icon={<InfoCircleIcon />} onClose={Function.prototype} isTruncated render={({
      className,
      content
    }) => (
      <FakeRouterLink to="somewhere" className={className}>
        {content}
      </FakeRouterLink>
    )}>
      Blue label fake router link with icon that overflows
    </Label>
```

and in code
```
export interface LabelProps {
  ...
  // haven't verified accuracy of the type definition
  render?: ({ className: string, content: React.ReactNode } => React.ReactNode)
}

export const Label = ({...props}) => {
  const content = (
    <>
      {icon && <span className={css(styles.labelIcon)}>{icon}</span>}
      {isTruncated && <span className={css(styles.labelText)}>{children}</span>}
      {!isTruncated && children}
    </>
  );
  return (
    <LabelComponent
      {...props}
      className={css(
        styles.label,
        colorStyles[color],
        variant === 'outline' && styles.modifiers.outline,
        isOverflowLabel && styles.modifiers.overflow,
        className
      )}
    >
      {render ? (
        render({
          className: styles.labelContent,
          content
        })
      ) : (
        <Component className={css(styles.labelContent)} {...(href && { href })}>
          {content}
        </Component>
      )}
      {onClose && button}
    </LabelComponent>
  );
}
```

File: packages/react-core/src/components/Label/Label.tsx (hunk 1 above)
Comment (jschuler) [reply packages/react-core/src/components/Label/Label.tsx:96]: yes if the code is changed to clone the child and apply our classname (not recommended) or better to use renderprops and pass that classname back to the user so they can apply it (see my example below)

**Comment by tlabaj** [comment]
> @tlabaj I would use the render props pattern here. Basically it's just another prop (usually called `render`) that gives the user control over the returned content. I could see it being used like this:
> 
> ```
> const FakeRouterLink = ({ to, children, ...props }) => <a href={to} {...props}>{children}</a>;
> ...
> <Label color="blue" icon={<InfoCircleIcon />} onClose={Function.prototype} isTruncated render={({
>       className, // this is the 'pf-c-label__content' class the user gets back
>       content // this is the content which could just be the string passed back to them, or enclosed in a text container (and icon sibling if applicable)
>     }) => (
>       <FakeRouterLink to="somewhere" className={className}>
>         {content}
>       </FakeRouterLink>
>     )}>
>       Blue label fake router link with icon that overflows
>     </Label>
> ```
> 
> and in code
> 
> ```
> export interface LabelProps {
>   ...
>   // haven't verified accuracy of the type definition
>   render?: ({ className, content }: { className: string, content: React.ReactNode } => React.ReactNode)
> }
> 
> export const Label = ({...props}) => {
>   const content = (
>     <>
>       {icon && <span className={css(styles.labelIcon)}>{icon}</span>}
>       {isTruncated && <span className={css(styles.labelText)}>{children}</span>}
>       {!isTruncated && children}
>     </>
>   );
>   return (
>     <LabelComponent
>       {...props}
>       className={css(
>         styles.label,
>         colorStyles[color],
>         variant === 'outline' && styles.modifiers.outline,
>         isOverflowLabel && styles.modifiers.overflow,
>         className
>       )}
>     >
>       {render ? (
>         render({
>           className: styles.labelContent,
>           content
>         })
>       ) : (
>         <Component className={css(styles.labelContent)} {...(href && { href })}>
>           {content}
>         </Component>
>       )}
>       {onClose && button}
>     </LabelComponent>
>   );
> }
> ```

@jschuler thanks for the quick feedback.  That is what I was looking for.  I will give it a try.

File: packages/react-core/src/components/Label/Label.tsx (hunk 1 above)
Comment (tlabaj) [reply packages/react-core/src/components/Label/Label.tsx:96]: @mcoker I am trying Joachim's suggestion below.

File: packages/react-core/src/components/Label/examples/Label.md [hunk 2]
```diff
@@ -6,6 +6,7 @@ propComponents: ['Label']
 ---
 
 import InfoCircleIcon from '@patternfly/react-icons/dist/js/icons/info-circle-icon';
+import { Link } from '@reach/router';
```
Comment (jschuler) [inline packages/react-core/src/components/Label/examples/Label.md:9]: we don't actually have this in our dependencies, maybe better to create a fake component as an example
`const FakeRouterLink = ({ to, children, ...props }) => <a href={to} {...props}>{children}</a>;`

File: packages/react-core/src/components/Label/examples/Label.md [hunk 3]
```diff
@@ -329,3 +330,24 @@ OutlinedLabels = () => (
   </React.Fragment>
 );
 ```
+
+### Router link
+```js
+import React from 'react';
+import { Label } from '@patternfly/react-core';
+import InfoCircleIcon from '@patternfly/react-icons/dist/js/icons/info-circle-icon';
+import { Link } from '@reach/router';
```
Comment (jschuler) [inline packages/react-core/src/components/Label/examples/Label.md:339]: same comment as above

**Review by mcoker — APPROVED** [review]
Looks great!

File: packages/react-core/src/components/Label/examples/Label.md (hunk 2 above)
Comment (jschuler) [reply packages/react-core/src/components/Label/examples/Label.md:9]: well somehow it works, not exactly sure how @redallen. I guess you can leave it!

Comment (redallen) [reply packages/react-core/src/components/Label/examples/Label.md:9]: It works because the theme uses that router and it's globally available there.

**Review by redallen — APPROVED** [review]
LGTM!

File: packages/react-core/src/components/Label/examples/Label.md (hunk 3 above)
Comment (tlabaj) [reply packages/react-core/src/components/Label/examples/Label.md:339]: Yes.  I saw it used in Dropdown and that works... I assumed it was in the docs?

**Review by jschuler — APPROVED** [review]

**Pull request merged**